bus-mapping/src/circuit_input_builder/input_state_ref.rs

```diff
@@ -986,25 +996,49 @@ impl<'a> CircuitInputStateRef<'a> {
     }
 
     /// Bus mapping for the RestoreContextGadget as used in RETURN.
-    // TODO: unify this with restore context bus mapping for STOP.
-    // TODO: unify this with the `handle return function above.`
     pub fn handle_restore_context(
         &mut self,
-        steps: &[GethExecStep],
         exec_step: &mut ExecStep,
+        steps: &[GethExecStep],
     ) -> Result<(), Error> {
         let call = self.call()?.clone();
+        let geth_step = &steps[0];
+        let is_return_revert = geth_step.op == OpcodeId::REVERT || geth_step.op == OpcodeId::RETURN;
+
+        if !is_return_revert && !call.is_success {
```

Comment:
Thanks! I do miss this, https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/evm/opcodes/return_revert.rs#L36-L41 
```
    self.call_context_read(
                exec_step,
                call.call_id,
                CallContextField::IsSuccess,
                call.is_success.to_word(),
    );
```


After I look deeper into the code, I couln't find a better way to resolve it. There are special cases basically,
1. A special case for gerenal error cases, https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1342/files#diff-29986cad7a3020e921e47cd1d6f7a667e1a06badff428e47686d8d038d845ad3R334-R346. It's hard to avoid this special case without any flags. And I don't think it's a good idea to move this piece of code into `handle_restore_context` which makes it more compilcated.
2. non-error opcode handler. The challenging is `return_revert`, https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3540a248efad5784d673e0d4336d89ed5569f875/bus-mapping/src/evm/opcodes/return_revert.rs#L93-L96. We can't just move this snippet into `handle_restore_context` bcs the order of `rw_table`. Ofc we could change the order in `evm-circuit`, but the way it's implemented now is much clear for me (it tells it's for `Case C in the specs.`)

So, based on above reasons, I think it's better to keep what it is now. If you find a better way to solve above issues, feel free to let me know and I could refine it!

Reply:
understood 👍 other LGTM!